Fixes #13216 - Added ability to extend API by creating facet subnode

ShimShtein opened the pull request:
This is a second part of host facets change. It contains a code that enables access to registered facets from the host model and API (v2) extension that enables adding properties to host's JSON.

First part of the change was merged here: #2944.
More complete explanations can be found here: #2665.

ShimShtein committed:
Fixes #13216 - Added ability to extend API by creating facet subnode
This commit adds the ability for plugins to extend Host model with custom data that will be stored in a facet model.
In order to show this data an extension point is added to Host's view API, so the properties list can be extended by the facet.
Parts that are contained in this PR:
1. Added extension point to API V2: Now each facet will be able to add facet's properties to host's details
2. Added model extensions, now you can enumerate all active facets using Host#facets and Host#facets_with_definitions methods
3. Host#attributes will include the facets by default
4. Includes facets in Host#clone

ShimShtein commented:
Rubocop issue fixed

ohadlevy commented:
@ShimShtein can you please beef up your commit message, I'm unable to understand what you are trying to do in this patch :)

ShimShtein commented:
1. Added extension point to API V2: Now each facet will be able to add facet's properties to host's details
2. Added model extensions, now you can enumerate all active facets using `Host#facets` and `Host#facets_with_definitions` methods
3. `Host#attributes` will include the facets by default
4. Includes facets in `Host#clone`

jlsherrill commented:
@ShimShtein I was under the impression you were going to open these up one by one rather than adding all 4 into one pr?

jlsherrill commented:
Tested the rabl extension points and works quite well.  No namespacing issues this time around :+1: 

Here's the work so far (although not 100% complete):  https://github.com/jlsherrill/katello/commit/cc54e3c3b9814a7df94f993f592cba9298199e6c

One odd thing i am seeing is that the registered facets seem to be disappearing from memory in  development.  Its unrelated to this PR but i just wanted to mention.  Likely related to some sort of class reloading rails is doing?

ShimShtein commented:
> I was under the impression you were going to open these up one by one rather than adding all 4 into one pr?

All extension points to the model and UI are omitted here. They will be released later, before puppet facet separation.

> registered facets seem to be disappearing from memory in development

I had this bug too. Wrap your initializer in `Rails.application.config.to_prepare do` and Rails will maintain the registration on reloading. The problem is that rails reloads the configuration module, so it needs to reload the configuration too.
@ares, I have heard that you had this kind of problem too, what do you think about this solution?

``` Ruby
Rails.application.config.to_prepare do
  Facets.register(PuppetFacet) do
    extend_model PuppetHostExtensions
    add_helper PuppetFacetHelper
    add_tabs :puppet_tabs
    api_view :list => 'api/v2/puppet_facets/base', :single => 'api/v2/puppet_facets/single_host_view'
    template_compatibility_properties :environment_id, :puppet_proxy_id, :puppet_ca_proxy_id
  end
end
```

ares commented:
If you load your classes in `config.to_prepare` block for the first time, it should help, rails autoloader will reload these objects and they would be re-registered as to_prepare is run before each request in development environment. If you'd use them before it would try to reload them but old ones would still be registered, resulting in various weird problems. So yes, sounds right to me.

dLobatog commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -75,6 +75,13 @@ def show
+
+          Facets.registered_facets.values.each do |facet_config|
+            next unless facet_config.api_param_group && facet_config.api_controller
+            param "#{facet_config.name}_attributes".to_sym, Hash, :desc => facet_config.api_param_group_description || N_("Parameters for host's %s facet" % facet_config.name) do
```
Translation syntax should be `(N_("Parameters for host's %s facet") % facet_config.name)`

dLobatog commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,66 @@
+      Facets::ManagedHostExtensions.refresh_facet_relations(self)
+
+      def attributes
+        hash = super
```
What is this hash, `host_attributes`?

ShimShtein replied:
Yes, it adds facets inside the attributes hash:

``` Ruby

host = Host.first
host.attributes
# => {
#   'id' => 1
#   'name' => 'myhost.example.com'
#   'puppet_facet_attributes' => {
#     'environment_id' => 1
#     }
# }
```

dLobatog commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,66 @@
+      end
+    end
+
+    def self.refresh_facet_relations(klass)
```
Nitpick, it's a bit more idiomatic to use `class << self` to avoid having to call `self.method` when you define several class methods, see https://github.com/bbatsov/ruby-style-guide#def-self-class-methods

dLobatog commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,66 @@
+      end
+    end
+
+    def self.register_facet_relation(klass, facet_config)
```
Please add some docs on this method, like "Defines relation between hosts and facets ...."

dLobatog commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,66 @@
+
+        include_in_clone facet_config.name
+
+        facet_config.compatibility_properties.each do |prop|
```
How about extracting to a separate method to be called from `refresh_facet_relations`? That way you can add docs, override it if needed, etc.. Explain what compatibility_properties are

ShimShtein replied:
The usage is documented here: https://github.com/theforeman/foreman/blob/develop/app/services/facets/entry.rb#L67

I will add some development documentation to the whole `register_facet_relation` method and explain more about compatibility properties there.

dLobatog commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,66 @@
+    # This method will return a hash of facets for a specific host including the coresponding definitions.
+    # The output should look like this:
+    # { host.puppet_aspect => Facets.registered_facets[:puppet_aspect] }
+    def facets_with_definitions
```
Nitpick, if what you're returning is the `facet_config`, why calling it `facets_with_definitions`? Maybe calling this just `facets`, and renaming `facet_names` above would work better. 

In any case when one calls `.something_in_plural` in Rails we normally expect it to be an Array of relations (i.e: `host.environments` would return an array of environments, not a hash of `environment.name => environment`). How about then returning an array of facets?

dLobatog replied:
Method is untested too

ShimShtein replied:
`#facets` returns an array of active records - the actual facet objects, not names.
`#facets_with_definitions` returns tuples: facet record as a key for the hash, same facet's configuration object as a value. Will it be better if the return value would be array of arrays?
Something like this: 

``` Ruby
[[facet1_instance, facet1_config], [facet2_instance, facet2_config]]
```

Or maybe the better solution would be just renaming the method to `#facet_and_cofigurations_hash`?

dLobatog replied:
That's fine, I got it wrong. I had some `pry` configuration so that the keys for facets_with_definitions would look like strings.

dLobatog commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,66 @@
+
+    def self.refresh_facet_relations(klass)
+      Facets.registered_facets.values.each do |facet_config|
+        self.register_facet_relation(klass, facet_config)
```
Untested

ShimShtein replied:
Both `#register_facet_relation` and `#refresh_facet_relations` are tested here: https://github.com/theforeman/foreman/pull/3076/files#diff-73245f1db9232d228a0ba8a982109379R26

They both are implementation details. Every time a facet registers, the registration work is done by `#register_facet_relation` method. Every time HostManagedExtensions class is included in Host, `#refresh` method will be called.

dLobatog replied:
Fair enough, thanks

dLobatog commented:
@ShimShtein Nice job, I don't have many objections - let me list them so it's easier to keep track of them:
- refresh_facet_relations is untested
- facets_with_definitions is untested
- API translation syntax is wrong
- How about renaming `facets_with_definitions` to `facet_configs`?
- How about returning an array with all facets relations when calling `host.facets` instead of names? 
- register_facet_relations and facets.compatibility_properties should be a bit more documented
- Use class << self on facets managed_hosts_extensions
- Please add more info, background and links to the commit message. I don't think someone without background on what facets do could understand what is the goal for this commit.

Let me know what you think, thanks!

ShimShtein commented:
> refresh_facet_relations is untested
> facets_with_definitions is untested

[answered](https://github.com/theforeman/foreman/pull/3076#discussion_r50124016)

> API translation syntax is wrong
> register_facet_relations and facets.compatibility_properties should be a bit more documented
> Use class << self on facets managed_hosts_extensions

Fixed

> How about renaming facets_with_definitions to facet_configs?
> How about returning an array with all facets relations when calling host.facets instead of names?

[answered](https://github.com/theforeman/foreman/pull/3076#discussion_r50118254)

> Please add more info, background and links to the commit message. I don't think someone without background on what facets do could understand what is the goal for this commit.

Done.

dLobatog commented:
[test]

dLobatog commented:
@ShimShtein :+1: if tests pass

dLobatog commented:
Usual integration tests failing - merging... http://ci.theforeman.org/job/test_develop_pr_core/7821/

dLobatog commented:
Merged as c804d530bcbe473f26187ebc800d206feb3c50f0, thanks @ShimShtein!

domcleal commented:
The commit message states this is a plugin API, so please add it to the wiki here, there's no documentation on facets at the moment: http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin.  It should also be exposed through the plugin registration API.
